assigning result of Bool_val to bool generates error (when used in C++ context)

```
stubs.c:561:54: fatal error: non-constant-expression cannot be narrowed from type 'int' to 'bool' in initializer list [-Wc++11-narrowing]
  struct bpf_filter f { bytes_of_value(Field(v, 0)), Bool_val(Field(v, 1)), t };
                                                     ^~~~~~~~~~~~~~~~~~~~~
/home/user/code/monorepo/_opam/lib/ocaml/caml/mlvalues.h:388:21: note: expanded from macro 'Bool_val'
#define Bool_val(x) Int_val(x)
                    ^~~~~~~~~~
/home/user/code/monorepo/_opam/lib/ocaml/caml/mlvalues.h:80:20: note: expanded from macro 'Int_val'
#define Int_val(x) ((int) Long_val(x))
```

~it was a warning with gcc, but new-ish~ clang makes it an error. It is fixable on the "user" side with an explicit cast but it looks ugly (`(bool)Bool_val`). Does it make sense to have this cast in mlvalues.h guarded with `#ifdef __cplusplus`?

--- a/Changes
+++ b/Changes
@@ -71,6 +71,9 @@ Working version
   of `caml_read_directory`.  Also, check for overflows in ext table sizes.
   (Xavier Leroy, report by Arseniy Alekseyev, review by Gabriel Scherer)
 
+- #11332, #12702: make sure `Bool_val(v)` has type `bool` in C++
+  (Xavier Leroy, report by ygrek, review by Gabriel Scherer)
+
 - #12223: Constify constructors and flags tables in C code. Now these
   tables will go in the readonly segment, where they belong.
   (Antonin Décimo, review by Gabriel Scherer and Xavier Leroy)

--- a/runtime/caml/mlvalues.h
+++ b/runtime/caml/mlvalues.h
@@ -425,7 +425,11 @@ CAMLextern value caml_atom(tag_t);
 /* Booleans are integers 0 or 1 */
 
 #define Val_bool(x) Val_int((x) != 0)
+#ifdef __cplusplus
+#define Bool_val(x) ((bool) Int_val(x))
+#else
 #define Bool_val(x) Int_val(x)
+#endif
 #define Val_false Val_int(0)
 #define Val_true Val_int(1)
 #define Val_not(x) (Val_false + Val_true - (x))

